Add sender constrained tokens (redo) (#95)
oauth-wg/oauth-identity-chaining: draft-ietf-oauth-identity-chaining.md

## Overview

> The identity and authorization chaining flow outlined below describes how a combination of OAuth 2.0 Token Exchange {{RFC8693}} and JWT Profile for OAuth 2.0 Client Authentication and Authorization Grants {{RFC7523}} are used to address the use cases identified. Section [Examples](#examples) include two additional examples that describe how this flow is used. In one example, the resource server acts as the client and in the other, the authorization server acts as the client.

Reviewer: https://drafts.oauth.net/oauth-identity-chaining/kelley/add-sender-constraints/draft-ietf-oauth-identity-chaining.html#name-overview this currently shows up as 

> Section [Appendix B](https://drafts.oauth.net/oauth-identity-chaining/kelley/add-sender-constraints/draft-ietf-oauth-identity-chaining.html#examples) includes two additional

Reviewer: Actually that's the preview for a different branch and I don't see one for this branch and I don't know why 

https://drafts.oauth.net/oauth-identity-chaining/

Reviewer: oh, it's because the "build" for this PR failed 

https://github.com/oauth-wg/oauth-identity-chaining/actions/runs/10864604120/job/30150062041?pr=95

```
...
/github/workspace/draft-ietf-oauth-identity-chaining.xml(654): Warning: Artwork too wide, reducing indentation from 0 to 0
/github/workspace/draft-ietf-oauth-identity-chaining.xml(652): Warning: Figure too wide, reducing indentation from 3 to 1
/github/workspace/draft-ietf-oauth-identity-chaining.xml(716): Warning: Artwork too wide, reducing indentation from 0 to 0
/github/workspace/draft-ietf-oauth-identity-chaining.xml(714): Warning: Figure too wide, reducing indentation from 3 to 1
/github/workspace/draft-ietf-oauth-identity-chaining.xml(232): Warning: Too long line found (L361), 3 characters longer than 72 characters: 
&subject_token_type=urn%3Aietf%3Aparams%3Aoauth%3Atoken-type%3Aaccess_token
lib/trace.sh draft-ietf-oauth-identity-chaining.html -s xml2rfc-html xml2rfc -q --rfc-base-url https://www.rfc-editor.org/rfc/ --id-base-url https://datatracker.ietf.org/doc/html/ --cache=/github/workspace/.refcache --html --css=lib/v3.css --metadata-js-url=/dev/null draft-ietf-oauth-identity-chaining.xml -o draft-ietf-oauth-identity-chaining.html
/github/workspace/draft-ietf-oauth-identity-chaining.xml(13): Warning: Expected a title or title abbreviation of not more than 40 character for the page header, found 56 characters
147:requested_cnf_supported_mtls 
188:  * AS-A MUST verify the proof of possession in the token exchange request. 
485:  
draft-ietf-oauth-identity-chaining.md contains trailing whitespace
*** Run 'make fix-lint' to automatically fix some errors
make[1]: *** [lib/main.mk:202: lint-whitespace] Error 1
make[1]: Target 'lint' not remade because of errors.
rm draft-ietf-oauth-identity-chaining.xml
make: *** [lib/main.mk:16: all] Error 2
```

Reviewer: that makes it hard to review

Author: I'm struggling with the web interface that I'm stuck using. Will continue working to resolve

Author: @bc-pi Pieter figured it out. Take a look now

Reviewer: That part doesn't look like it renders like you'd want 
https://drafts.oauth.net/oauth-identity-chaining/add-sender-constrained-tokens/draft-ietf-oauth-identity-chaining.html#name-overview

Author: Not sure what you mean. Let's talk next week

Reviewer: Maybe just:
```suggestion
The identity and authorization chaining flow outlined below describes how a combination of OAuth 2.0 Token Exchange {{RFC8693}} and JWT Profile for OAuth 2.0 Client Authentication and Authorization Grants {{RFC7523}} are used to address the use cases identified. (#examples) include two additional examples that describe how this flow is used. In one example, the resource server acts as the client and in the other, the authorization server acts as the client.
```

## Authorization Server Discovery
This specification does not define authorization server discovery. A client MAY maintain a static mapping or use other means to identify the authorization server. The `authorization_servers` property in {{I-D.ietf-oauth-resource-metadata}} MAY be used.

## Sender Constrained Tokens
Either authorization server MAY issue sender constrained tokens. There are currently two options for sender constrained tokens: Mutually Authenticated TLS (mTLS) {{RFC8705}} and Demonstrating Proof of Possessions (DPoP) {{RFC9449}}.

The following authorization server metadata value is used by this specification when the authorization server is acting as the Client.

{

>    "requested_conf_supported": true

Author: requested_cnf_supported_dpop
requested_cnf_supported_mtls

This document introduces the following authorization server metadata values to indicate the authorization server's support for DPoP or mTLS sender constrained tokens.

requested_cnf_supported_dpop
: OPTIONAL. Boolean indicator of whether the authorization server supports DPoP sender constrained tokens.
When the authorization server is acting as the Client, the Client MAY check whether AS-B supports sender constrained tokens by obtaining the authorization server metadata and verifying "requested_conf_supported" is set to true.

requested_cnf_supported_mtls 
: OPTIONAL. Boolean indicator of whether the authorization server supports mTLS sender constrained tokens.
When the authorization server is acting as the Client, the Client MAY check whether AS-B supports sender constrained tokens by obtaining the authorization server metadata and verifying the appropriate sender constrained token method is set to true.

When the authorization server is acting as the Client, the Client MAY check whether AS-B supports sender constrained tokens by obtaining the authorization server metadata and verifying the appropriate sender constrained token method is set to true.

> If the inbound token is sender constrained, RS-A MUST verify the proof of possession for the inbound token before proceeding to the Token Exchange Request.

Reviewer: I'm not sure about this whole section TBH for a number of reasons that escape my ability to articulate easily or meaningfully.

## Authorization Server Discovery
This specification does not define authorization server discovery. A client MAY maintain a static mapping or use other means to identify the authorization server. The `authorization_servers` property in {{I-D.ietf-oauth-resource-metadata}} MAY be used.

## Sender Constrained Tokens
Either authorization server MAY issue sender constrained tokens. There are currently two options for sender constrained tokens: Mutually Authenticated TLS (mTLS) {{RFC8705}} and Demonstrating Proof of Possessions (DPoP) {{RFC9449}}.

> This document introduces the following authorization server metadata values to indicate the authorization server's support for DPoP or mTLS sender constrained tokens.

Reviewer: I think "authorization server's support for DPoP or mTLS sender constrained tokens" is much too general

Reviewer: similar thoughts about below text too

## Authorization Server Discovery
This specification does not define authorization server discovery. A client MAY maintain a static mapping or use other means to identify the authorization server. The `authorization_servers` property in {{I-D.ietf-oauth-resource-metadata}} MAY be used.

## Sender Constrained Tokens

> Either authorization server MAY issue sender constrained tokens. There are currently two options for sender constrained tokens: Mutually Authenticated TLS (mTLS) {{RFC8705}} and Demonstrating Proof of Possessions (DPoP) {{RFC9449}}.

Reviewer 2: Readers may think this includes the assertion token. I believe limiting this to access tokens makes it more clear.
```suggestion
Either authorization server MAY issue sender constrained access tokens. There are currently two options for sender constrained tokens: Mutually Authenticated TLS (mTLS) {{RFC8705}} and Demonstrating Proof of Possessions (DPoP) {{RFC9449}}.
```

Reviewer: also Possession is singular 
```suggestion
Either authorization server MAY issue sender constrained access tokens. There are currently two options for sender constrained tokens: Mutually Authenticated TLS (mTLS) {{RFC8705}} and Demonstrating Proof of Possession (DPoP) {{RFC9449}}.
```

* If the request itself is not valid or if the given resource or audience are unknown, or are unacceptable based on policy, the authorization server MUST deny the request.
* The authorization server MAY add, remove or change claims. See [Claims transcription](#claims-transcription).

> * If the inbound token is sender constrained and the resource server is acting as the Client, the Client MUST provide proof of possession of a private key, using either mTLS or DPoP, in the token exchange request to AS-A in Step (B). This is not applicable to the authorization server acting as the Client since it performs an "internal token exchange".

Author: "of its private key.."
Add to security considerations: why?: for consistency of end-to-end sender constrained tokens
"Step (B)" to maintain the practice of sender constraining tokens throughout, or something like that

* The "aud" claim included in the returned JWT authorization grant MAY identify multiple authorization servers, provided that trust relationships exist with them (e.g. through federation). It is RECOMMENDED that the "aud" claim is restricted to a single authorization server to prevent an authorization server in one domain from presenting the client's authorization grant to an authorization server in a different trust domain. For example, this will prevent the authorization server in Domain B from presenting the authorization grant it received from the client in Domain A to the authorization server for Domain C.

* If the token exchange request in (B) contains a proof of possession of a private key, the following apply.

>   * AS-A MUST verify the proof of possession for the token in the request (the initial inbound token)

Reviewer 2: I'm not sure how AS-A would be able to verify the proof of possession:
- If mTLS is used the connecting peer of AS-A is RS-A, which key would differ to the one present in the cnf claim ("clients" key in "cnf" vs "RS" key in mTLS)
- If DPoP the proof could be "forwarded" but how would AS-A know what audience, htm, htu to expect? And they'll differ to the endpoint and method used during token exchange.

Reviewer 2: I only realized during the meeting that the `cnf` key of the resulting access token will not be the same as the `cnf` key of the incoming access tokens.
Maybe it would be good to mention this in the draft. E.g. 

```
* If the JWT authorization grant in the access token request contains a "requested_cnf" claim **with the key of the client performing the token exchange. This will bind the access token for domain B to the client planning to use it.**
```

* If the request itself is not valid or if the given resource or audience are unknown, or are unacceptable based on policy, the authorization server MUST deny the request.
* The authorization server MAY add, remove or change claims. See [Claims transcription](#claims-transcription).

> * If the inbound token is sender constrained and the resource server is acting as the Client, the Client MUST provide proof of possession of its private key, using either mTLS or DPoP, in the token exchange request to AS-A in Step (B). This is not applicable to the authorization server acting as the Client since it performs an "internal token exchange".

Reviewer: I'm admittedly a bit confused at present but I think the "the inbound token is sender constrained and the resource server is acting as the Client, the Client MUST provide proof of possession of its private key, using either mTLS or DPoP, in the token exchange request to AS-A in Step (B)" part isn't possible?

* The "aud" claim included in the returned JWT authorization grant MAY identify multiple authorization servers, provided that trust relationships exist with them (e.g. through federation). It is RECOMMENDED that the "aud" claim is restricted to a single authorization server to prevent an authorization server in one domain from presenting the client's authorization grant to an authorization server in a different trust domain. For example, this will prevent the authorization server in Domain B from presenting the authorization grant it received from the client in Domain A to the authorization server for Domain C.

* If the token exchange request in (B) contains a proof of possession of the Client's private key, the following apply.
  * AS-A MUST verify the proof of possession in the token exchange request. 

>   * AS-A MUST include the “cnf” claim from the “requested_cnf” claim in the token exchange request in the returned authorization grant.

Reviewer: I suspect that the smart quotes on `“cnf”` and `“requested_cnf”` and elsewhere might be cause of some issues. Maybe. 

Also a new claim like `requested_cnf` needs to be defined somewhere rather than just casually introduced. And also request to be registered but let's ignore that for the time being.

The client MAY indicate the audience it is trying to access through the `scope` parameter or the `resource` parameter defined in {{RFC8707}}.

> If the authorization server is acting as the Client, the Client MUST include the “cnf” claim from the token exchange request in a “requested_cnf” claim in the returned authorization grant. This does not apply when the resource server is acting as the Client.

Reviewer: wait... what?

The authorization server responds with an access token as described in section 5.1 of {{RFC6749}}.

The following process rule applies:

> * If the JWT authorization grant in the access token request contains a "requested_cnf" claim, the authorization server MUST include the "cnf" claim from the "requested_cnf" claim in the returned token.

Reviewer 2: If AS of domain B does not support sender constraining tokens it should still be able to issue an access token (but without 'cnf'). Otherwise, chaining from sender constraint domain to non-constraint domain won't be possible.

> * If the JWT authorization grant in the access token request contains a "requested_cnf" claim, the authorization server MUST include the "cnf" claim from the "requested_cnf" claim in the returned token.

Reviewer 2: Pieter suggested to add a metadata field to authorization server metadata that indicates if sender constraint tokens are supported or not. 

@PieterKas would that field be general about "cnf" is supported or should it rather be "assertion with requested_cnf is supported"?

The authorization server responds with an access token as described in section 5.1 of {{RFC6749}}.

The following process rule applies:

* If the JWT authorization grant in the access token request contains a "requested_cnf" claim, the authorization server MUST include the "cnf" claim from the "requested_cnf" claim in the returned token.
* If Domain B does not support sender constrained tokens and the request is sender constrained, AS-B MAY issue access tokens that are not sender constrained.

> * If Domain B supports sender constrained tokens and the request is not sender constrained, AS-B MUST reject the request and return an error response per Section 5.2 of {{RFC6749}} with invalid_pop as the value of the error parameter.

Author: supports -> requires

The following process rule applies when the authorization server is acting as the Client

> * If Domain B supports sender constrained tokens and the JWT authorization grant in the access token request contains a "requested_cnf" claim, the authorization server MUST include the "cnf" claim from the "requested_cnf" claim in the returned token.

Author: AS MUST verify the pop for the private key corresponding to the cnf claim

The authorization server responds with an access token as described in section 5.1 of {{RFC6749}}.

The following processing rules applies:

* If the JWT authorization grant in the Access Token Request contains a "requested_cnf" claim:

>   * The authorization server in Domain B MUST verify the proof of possession for the private key corresponding to the top level "cnf" claim.

Reviewer: how?

  * The authorization server MUST include the "cnf" value from the "requested_cnf" claim in the returned token.

> * If Domain B does not support sender constrained tokens and the rAccess Token Request is sender constrained, AS-B MAY issue access tokens that are not sender constrained.

Reviewer: ```suggestion
* If Domain B does not support sender constrained tokens and the Access Token Request is sender constrained, AS-B MAY issue access tokens that are not sender constrained.
```

Reviewer: but what is a sender constrained Access Token Request?

> * If Domain B requires constrained tokens and the Token Exchange Rquest is not sender constrained, AS-B MUST reject the request and return an error response per Section 5.2 of {{RFC6749}} with invalid_pop as the value of the error parameter.

Reviewer: would need to request registration of a new error code...

Reviewer: and we should really try to be consistent with how protocol elements are quoted or ticked or what. Or at least try not to more inconsistent :)
